tools/urtrace/collector.cpp

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * @file collector.cpp
+ *
+ * This file contains the implementation of the UR collector library. The library
+ * provides instrumentation for tracing function calls and profiling function
+ * execution time.
+ */
+
+#include <chrono>
+#include <optional>
+#include <regex>
+#include <sstream>
+#include <stack>
+#include <string>
+#include <string_view>
+
+#include "logger/ur_logger.hpp"
+#include "ur_api.h"
+#include "ur_params.hpp"
+#include "xpti/xpti_trace_framework.h"
+
+constexpr uint16_t TRACE_FN_BEGIN =
+    static_cast<uint16_t>(xpti::trace_point_type_t::function_with_args_begin);
+constexpr uint16_t TRACE_FN_END =
+    static_cast<uint16_t>(xpti::trace_point_type_t::function_with_args_end);
+constexpr std::string_view UR_STREAM_NAME = "ur";
+
+logger::Logger out = logger::create_logger("collector", true);
+
+#define ARGS_ENV "UR_COLLECTOR_ARGS"
+
+enum time_unit {
+    TIME_UNIT_AUTO,
+    TIME_UNIT_NS,
+    TIME_UNIT_US,
+    TIME_UNIT_MS,
+    TIME_UNIT_S,
+    MAX_TIME_UNIT,
+};
+
+const char *time_unit_str[MAX_TIME_UNIT] = {"auto", "ns", "us", "ms", "s"};
```

Comment:
Would making a `time_unit` and a `const char *time_unit_str` a map entry cause a performance hit here? It seems cleaner to use a map instead, however you won't be able to `switch` through all of the keys.

Reply:
I don't agree using a map is cleaner. This allocates no memory and is very straightforward. And it's an extremely common pattern for mapping enum values to strings...